src/papilo/io/PboParser.hpp

```diff
@@ -0,0 +1,397 @@
+/* along with this program.  If not, see <https://www.gnu.org/licenses/>.    */
+/*                                                                           */
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#ifndef _PAPILO_IO_PBO_PARSER_HPP_
+#define _PAPILO_IO_PBO_PARSER_HPP_
+
+#include "papilo/Config.hpp"
+#include "papilo/core/ConstraintMatrix.hpp"
+#include "papilo/core/Objective.hpp"
+#include "papilo/core/Problem.hpp"
+#include "papilo/core/VariableDomains.hpp"
+#include "papilo/external/pdqsort/pdqsort.h"
+#include "papilo/misc/Flags.hpp"
+#include "papilo/misc/Hash.hpp"
+#include "papilo/misc/Num.hpp"
+#include <algorithm>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/dynamic_bitset.hpp>
+#include <boost/iostreams/filtering_stream.hpp>
+#include <boost/optional.hpp>
+#include <boost/spirit/include/qi.hpp>
+#include <boost/utility/string_ref.hpp>
+#include <fstream>
+#include <iostream>
+#include <iterator>
+#include <limits>
+#include <memory>
+#include <tuple>
+#include <utility>
+
+#ifdef PAPILO_USE_BOOST_IOSTREAMS_WITH_BZIP2
+#include <boost/iostreams/filter/bzip2.hpp>
+#endif
+#ifdef PAPILO_USE_BOOST_IOSTREAMS_WITH_ZLIB
+#include <boost/iostreams/filter/gzip.hpp>
+#endif
+
+namespace papilo
+{
+
+template <typename REAL, bool isfptype = num_traits<REAL>::is_floating_point>
+struct RealParseType
+{
+   using type = double;
+};
+
+template <typename REAL>
+struct RealParseType<REAL, true>
+{
+   using type = REAL;
+};
+
+/// Parser for pbo files derived from MpsParser.hpp
+template <typename REAL>
+class PboParser
+{
+   static_assert(
+       num_traits<typename RealParseType<REAL>::type>::is_floating_point,
+       "the parse type must be a floating point type" );
+   // TODO replace this so fractional values are required instead.
+
+ public:
+   static boost::optional<Problem<REAL>>
+   loadProblem( const std::string& filename )
+   {
+      PboParser<REAL> parser;
+
+      Problem<REAL> problem;
+
+      if( !parser.parseFile( filename ) )
+         return boost::none;
+
+      assert( parser.nnz >= 0 );
+
+      Vec<REAL> obj_vec( size_t( parser.nCols ), REAL{ 0.0 } );
+
+      for( auto i : parser.coeffobj )
+         obj_vec[i.first] = i.second;
+
+      problem.setObjective( std::move( obj_vec ), parser.objoffset );
+      problem.setConstraintMatrix(
+          SparseStorage<REAL>{ std::move( parser.entries ), parser.nCols,
+                               parser.nRows, true },
+          std::move( parser.rowlhs ), std::move( parser.rowrhs ),
+          std::move( parser.row_flags ), true );
+      problem.setVariableDomains(
+          std::move( Vec<REAL> vect( n, REAL( 0 ) ) ),
+          std::move( Vec<REAL> vect( n, REAL( 1 ) ) ),
+          std::move( Vec<ColFlags> vect( n, kIntegral ) ) );
+      problem.setVariableNames( std::move( parser.colnames ) );
+      problem.setName( std::move( filename ) );
+      // We do not have ConstraintNames in PBO
```

Comment:
we definitely should give them names. 
Imho a simple counter would be the best solution.